pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,36 @@
+package kademlia
+
+import (
+	"context"
+
+	"storj.io/storj/protos/overlay"
+)
+
+type nodeID string
+
+// DHT is the interface for the DHT in the Storj network
+type DHT interface {
+	GetNodes(ctx context.Context) []overlay.Node
+	GetRoutingTable(ctx context.Context) RoutingTable
+	Bootstrap(ctx context.Context) error
+	Ping(ctx context.Context, node overlay.Node) overlay.Node
+	FindNode(ctx context.Context, ID nodeID) overlay.Node
+	FindValue(ctx context.Context)
+}
+
+// RoutingTable contains information on nodes we have locally
+type RoutingTable interface {
+	GetBuckets() []*Bucket
+	GetBucket(id string) (ok bool, bucket Bucket)
+}
+
+// Bucket is a set of methods to act on kademlia k buckets
+type Bucket interface {
+	MoveToTail(ID nodeID) error
+	MoveToHead(ID nodeID) error
+	Has(ID nodeID) bool
+	Add(ID nodeID) error
+	Remove(ID nodeID) error
+	Get(ID nodeID) overlay.Node
+	Len() int
```

Comment:
with the methods for the routing table above, the stuff you need in the Bucket are very little, maybe just 

```
type Bucket interface {
  Routing() []overlay.Node
  Cache() []overlay.Node
  Midpoint() string
}
```

Reply:
I was basing a lot of those off the paper where it describes adding/removing elements from a bucket. You thinking those shouldn't be exposed methods ?

Reply:
well, the only time elements are added or removed are from these two and only these two callback methods:

```
  ConnectionSuccess(id string, address overlay.NodeAddress)
  ConnectionFailed(id string, address overlay.NodeAddress)
```

so i think it's probably less error prone to have the ability to manipulate the bucket in just those two places? i dunno. 

maybe we need a high level interface and a low level interface, and the high level is implemented in terms of the low level?

Reply:
OK I think I grasp what you are saying. 

The logic of dropping an address, adding an address, splitting the buckets etc. will all be implemented directly within these two methods.